arch/risc-v/src/qemu-rv/qemu_rv_start.c

```diff
@@ -144,13 +146,22 @@ void qemu_rv_start(int mhartid)
 
 #ifdef CONFIG_BUILD_KERNEL
 
+/****************************************************************************
+ * Name: qemu_rv_start_s
+ ****************************************************************************/
+
 void qemu_rv_start_s(int mhartid)
 {
-  qemu_rv_clear_bss();
+  /* NOTE: still in M-mode */
 
-  /* Initialize the per CPU areas */
+  if (0 == mhartid)
+    {
+      qemu_rv_clear_bss();
+
+      /* Initialize the per CPU areas */
 
-  riscv_percpu_add_hart(mhartid);
+      riscv_percpu_add_hart(mhartid);
```

Comment:
should we do riscv_percpu_add_hart for all core not only he first one

Reply:
@xiaoxiang781216 
Because the per CPU areas exist in BSS and the areas are cleared by CPU0, so non-primary CPUs should call `riscv_percpu_add_hart(mhartid)` after the IPI received in `riscv_cpu_boot()` .

Reply:
In this case, should we let the secondary core wait the IPI before running any useful code? So, all many places can share the code path for primary and secondary core except one time initialization which just need be done for primary core.

Reply:
@xiaoxiang781216 
In RISC-V with BUILD_KERNEL case, IPI send/receive sequence can be done in S-mode.
However, some code initialization must be done in M-mode on each CPU as well before entering S-mode.

Reply:
Can we do the initialization in assembly code and then wait the IPI before enter the C?

Reply:
@xiaoxiang781216 
Let me consider it later but I think we should do that in another PR.
What do you think?

Reply:
sure